### PR TITLE
Add more impls to Pattern

### PR DESCRIPTION
**Pull request by sffc**
Depends on #4657
Depends on #4659
Depends on #4663

This adds Databake, Yokeable, and ZeroFrom. Serde will follow.

### CODE REVIEW

File: utils/pattern/src/frontend/databake.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    B: 'static,
+    Store: Bake,
+{
+    fn bake(&self, ctx: &CrateEnv) -> TokenStream {
+        ctx.insert("icu_pattern");
+        let store = self.store.bake(ctx);
+        let b = if TypeId::of::<B>() == TypeId::of::<SinglePlaceholder>() {
+            quote!(icu_pattern::SinglePlaceholder)
+        } else {
+            unreachable!("sealed")
```
Comment (Manishearth) [inline utils/pattern/src/frontend/databake.rs:23]: nit: a little bit more text here please

File: utils/pattern/src/frontend/mod.rs [hunk 2]
```diff
@@ -51,6 +60,17 @@ impl<Backend, Store> Pattern<Backend, Store> {
     pub fn take_store(self) -> Store {
         self.store
     }
+
+    /// # Safety
+    ///
+    /// The store must come from a valid `Pattern` with this `Backend`,
+    /// such as by calling [`Pattern::take_store()`].
+    pub const unsafe fn from_store_unchecked(store: Store) -> Self {
```
Comment (Manishearth) [inline utils/pattern/src/frontend/mod.rs:68]: question/issue: is this a memory safety requirement? if not, this shouldn't be unsafe

**Review by Manishearth — COMMENTED** [review]
pattern code is mostly fine

**Review by robertbastian — DISMISSED** [review]
changelog for databake please

**Review by Manishearth — APPROVED** [review]

**Comment by sffc** [comment]
> changelog for databake please

https://github.com/unicode-org/icu4x/pull/4663

**Review by Manishearth — APPROVED** [review]

**Pull request merged**